File: sdk/node/test/e2e/ingest.e2e.test.ts
// Real end-to-end suite: exercises the published SDK against a LIVE attune
// server over HTTP. Skipped unless ATTUNE_E2E_BASE_URL + ATTUNE_E2E_API_KEY are
// set (the scripts/e2e.sh harness boots a server and sets them). Env-driven, so
// it also runs against any real deployment:
//
//   ATTUNE_E2E_BASE_URL=https://attune.example.com \
//   ATTUNE_E2E_API_KEY=fbk_live_... \
//   pnpm exec vitest run test/e2e
import { describe, expect, it } from 'vitest'
import { AttuneError, Client } from '../../src/index'

const baseURL = process.env.ATTUNE_E2E_BASE_URL
const apiKey = process.env.ATTUNE_E2E_API_KEY
const live = describe.skipIf(!baseURL || !apiKey)
// Narrowed for use inside the suite (only runs when both are set).
const BASE = baseURL ?? ''
const KEY = apiKey ?? ''

// Tag content so the harness can find these rows in Postgres afterwards.
export const E2E_MARKER = process.env.ATTUNE_E2E_MARKER ?? 'sdk-e2e'

live('ingest against a live attune server', () => {
  const client = () => new Client({ baseURL: BASE, apiKey: KEY })

  describe('happy paths', () => {
    it('ingests minimal { content } and returns a numeric id + pending', async () => {
      const res = await client().ingest({ content: `${E2E_MARKER} minimal` })
      expect(res.enrichmentStatus).toBe('pending')
      expect(res.id).toMatch(/^\d+$/) // int64 serialized as a numeric string
    })

    it('ingests all optional fields (source=web, sourceUser, pageUrl, sourceMeta)', async () => {
      const res = await client().ingest({
        content: `${E2E_MARKER} full-fields`,
        source: 'web',
        sourceUser: 'e2e-user-42',
        pageUrl: 'https://app.example.com/settings',
        sourceMeta: { plan: 'pro', tries: 3 },
      })
      expect(res.id).toMatch(/^\d+$/)
      expect(res.enrichmentStatus).toBe('pending')
    })

    it('assigns distinct, increasing ids to sequential ingests', async () => {
      const a = await client().ingest({ content: `${E2E_MARKER} seq-a` })
      const b = await client().ingest({ content: `${E2E_MARKER} seq-b` })
      // Compare as BigInt — id is an int64 string; Number() would lose precision.
      expect(BigInt(b.id) > BigInt(a.id)).toBe(true)
    })

    it('accepts unicode content', async () => {
      const res = await client().ingest({ content: `${E2E_MARKER} café Ångström 🎉 €` })
      expect(res.id).toMatch(/^\d+$/)
    })

    it('accepts content at the 5000-char cap', async () => {
      const res = await client().ingest({ content: 'a'.repeat(5000) })
      expect(res.id).toMatch(/^\d+$/)
    })

    it('accepts every documented source', async () => {
      for (const source of ['api', 'webhook', 'email', 'web', 'other'] as const) {
        const res = await client().ingest({ content: `${E2E_MARKER} src-${source}`, source })
        expect(res.id).toMatch(/^\d+$/)
      }
    })
  })

  describe('server validation → AttuneError (no retry)', () => {
    it('rejects empty content with VALIDATION / 400', async () => {
      await expect(client().ingest({ content: '' })).rejects.toMatchObject({
        name: 'AttuneError',
        code: 'VALIDATION',
        status: 400,
      })
    })

    it('rejects content over the 5000-char cap with VALIDATION / 400', async () => {
      await expect(client().ingest({ content: 'a'.repeat(5001) })).rejects.toMatchObject({
        code: 'VALIDATION',
        status: 400,
      })
    })

    it('rejects an unknown source with VALIDATION / 400', async () => {
      await expect(
        client().ingest({ content: `${E2E_MARKER} bad-source`, source: 'totally-bogus' }),
      ).rejects.toMatchObject({ code: 'VALIDATION', status: 400 })
    })
  })

  describe('auth', () => {
    it('rejects an unknown api key with UNAUTHORIZED / 401 and a requestId', async () => {
      const bad = new Client({ baseURL: BASE, apiKey: 'fbk_live_definitely_invalid' })
      const err = await bad.ingest({ content: 'x' }).catch((e: unknown) => e)
      expect(err).toBeInstanceOf(AttuneError)
      expect(err).toMatchObject({ code: 'UNAUTHORIZED', status: 401 })
      expect((err as AttuneError).requestId).toBeTruthy()
    })
  })

  describe('idempotency', () => {
    it('replaying the same Idempotency-Key returns the same id (no duplicate row)', async () => {
      const key = `e2e-idem-${E2E_MARKER}-replay`
      const c = client()
      const first = await c.ingest(
        { content: `${E2E_MARKER} idem-replay` },
        { idempotencyKey: key },
      )
      const second = await c.ingest(
        { content: `${E2E_MARKER} idem-replay` },
        { idempotencyKey: key },
      )
      expect(second.id).toBe(first.id)
    })

    it('N CONCURRENT ingests with the same key collapse to one row (same id)', async () => {
      const key = `e2e-idem-${E2E_MARKER}-concurrent`
      const c = client()
      const results = await Promise.all(
        Array.from({ length: 8 }, () =>
          c.ingest({ content: `${E2E_MARKER} idem-concurrent` }, { idempotencyKey: key }),
        ),
      )
      const ids = new Set(results.map((r) => r.id))
      expect(ids.size).toBe(1) // all 8 concurrent calls resolved to the same row
    })

    it('same key with a different body → 409 IDEMPOTENCY_CONFLICT', async () => {
      const key = `e2e-idem-${E2E_MARKER}-conflict`
      const c = client()
      await c.ingest({ content: `${E2E_MARKER} idem-A` }, { idempotencyKey: key })
      await expect(
        c.ingest({ content: `${E2E_MARKER} idem-B-different` }, { idempotencyKey: key }),
      ).rejects.toMatchObject({ code: 'IDEMPOTENCY_CONFLICT', status: 409 })
    })
  })

  describe('transport failures', () => {
    it('surfaces a connection failure as code NETWORK', async () => {
      const dead = new Client({ baseURL: 'http://127.0.0.1:1', apiKey: KEY, maxRetries: 0 })
      await expect(dead.ingest({ content: 'x' })).rejects.toMatchObject({ code: 'NETWORK' })
    })

    it('surfaces a per-attempt timeout as code TIMEOUT', async () => {
      const slow = new Client({ baseURL: BASE, apiKey: KEY, timeout: 1, maxRetries: 0 })
      await expect(slow.ingest({ content: 'x' })).rejects.toMatchObject({ code: 'TIMEOUT' })
    })

    it('honors a pre-aborted signal with code ABORTED', async () => {
      const controller = new AbortController()
      controller.abort()
      await expect(
        client().ingest({ content: 'x' }, { signal: controller.signal }),
      ).rejects.toMatchObject({ code: 'ABORTED' })
    })
  })
})

File: sdk/go/test/e2e/ingest_e2e_test.go
//go:build e2e

// Package e2e exercises the published Go SDK against a real, running attune
// server. It is gated behind the `e2e` build tag so `go test ./...` skips it;
// scripts/e2e.sh boots the server + Postgres and runs it with `-tags e2e`.
//
// Required env:
//
//	ATTUNE_E2E_BASE_URL  base URL of the live server (e.g. http://127.0.0.1:8097)
//	ATTUNE_E2E_API_KEY   an ingest:write key
//	ATTUNE_E2E_MARKER    unique content prefix so the harness can find rows in PG
package e2e

import (
	"context"
	"errors"
	"os"
	"sync"
	"testing"
	"time"

	attune "github.com/Phixsura/attune/sdk/go"
)

func newClient(t *testing.T) (*attune.Client, string) {
	t.Helper()
	base := os.Getenv("ATTUNE_E2E_BASE_URL")
	key := os.Getenv("ATTUNE_E2E_API_KEY")
	marker := os.Getenv("ATTUNE_E2E_MARKER")
	if base == "" || key == "" || marker == "" {
		t.Skip("ATTUNE_E2E_BASE_URL / ATTUNE_E2E_API_KEY / ATTUNE_E2E_MARKER not set")
	}
	c, err := attune.New(base, key, attune.WithTimeout(10*time.Second))
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	return c, marker
}

func TestE2EBasicIngest(t *testing.T) {
	c, marker := newClient(t)
	ctx := context.Background()
	for i := 0; i < 6; i++ {
		res, err := c.Ingest(ctx, attune.IngestInput{Content: marker + " basic"})
		if err != nil {
			t.Fatalf("basic ingest %d: %v", i, err)
		}
		if res.ID == "" {
			t.Fatalf("basic ingest %d: empty id", i)
		}
		if res.EnrichmentStatus != "pending" {
			t.Errorf("basic ingest %d: status = %q, want pending", i, res.EnrichmentStatus)
		}
	}
}

func TestE2EFullFieldsIngest(t *testing.T) {
	c, marker := newClient(t)
	res, err := c.Ingest(context.Background(), attune.IngestInput{
		Content:    marker + " full-fields",
		Source:     "web",
		SourceUser: "e2e-user-42",
		PageURL:    "https://app.example.com/settings",
		SourceMeta: map[string]any{"plan": "pro"},
	})
	if err != nil {
		t.Fatalf("full-fields ingest: %v", err)
	}
	if res.ID == "" {
		t.Fatal("full-fields ingest: empty id")
	}
}

func TestE2EIdempotencyReplayDedup(t *testing.T) {
	c, marker := newClient(t)
	ctx := context.Background()
	key := "replay-" + marker
	first, err := c.Ingest(ctx, attune.IngestInput{Content: marker + " idem-replay"}, attune.WithIdempotencyKey(key))
	if err != nil {
		t.Fatalf("first replay ingest: %v", err)
	}
	second, err := c.Ingest(ctx, attune.IngestInput{Content: marker + " idem-replay"}, attune.WithIdempotencyKey(key))
	if err != nil {
		t.Fatalf("second replay ingest: %v", err)
	}
	if first.ID != second.ID {
		t.Errorf("replay returned different ids: %q vs %q (dedup failed)", first.ID, second.ID)
	}
}

// TestE2EUnauthorizedBadKey verifies the SDK surfaces a real 401 from the live
// server (error-envelope parsing against the actual API, not a mock).
func TestE2EUnauthorizedBadKey(t *testing.T) {
	base := os.Getenv("ATTUNE_E2E_BASE_URL")
	if base == "" {
		t.Skip("ATTUNE_E2E_BASE_URL not set")
	}
	c, err := attune.New(base, "att_sk_definitely_not_a_real_key_000000")
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	_, err = c.Ingest(context.Background(), attune.IngestInput{Content: "should be rejected"})
	var ae *attune.AttuneError
	if !errors.As(err, &ae) {
		t.Fatalf("want *AttuneError, got %v", err)
	}
	if ae.Status != 401 || ae.Code != attune.CodeUnauthorized {
		t.Errorf("got status=%d code=%s, want 401 UNAUTHORIZED", ae.Status, ae.Code)
	}
}

// TestE2EValidationEmptyContent verifies a real 4xx validation error and that
// the requestId from the server's error envelope is parsed through.
func TestE2EValidationEmptyContent(t *testing.T) {
	c, _ := newClient(t)
	_, err := c.Ingest(context.Background(), attune.IngestInput{Content: ""})
	var ae *attune.AttuneError
	if !errors.As(err, &ae) {
		t.Fatalf("want *AttuneError, got %v", err)
	}
	if ae.Status != 400 {
		t.Errorf("got status=%d code=%s, want 400", ae.Status, ae.Code)
	}
	if ae.RequestID == "" {
		t.Error("expected a non-empty requestId parsed from the server error envelope")
	}
}

func TestE2EConcurrentDedup(t *testing.T) {
	c, marker := newClient(t)
	key := "concurrent-" + marker
	const n = 8
	ids := make([]string, n)
	errs := make([]error, n)
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			res, err := c.Ingest(context.Background(),
				attune.IngestInput{Content: marker + " idem-concurrent"},
				attune.WithIdempotencyKey(key))
			ids[i], errs[i] = res.ID, err
		}(i)
	}
	wg.Wait()
	for i, err := range errs {
		if err != nil {
			t.Fatalf("concurrent ingest %d: %v", i, err)
		}
	}
	for i := 1; i < n; i++ {
		if ids[i] != ids[0] {
			t.Errorf("concurrent ingest returned different ids: %q vs %q (dedup failed)", ids[0], ids[i])
		}
	}
}
